src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+    if (!cluster_idx.has_value())
+    {
+        return InteractionModel::ClusterEntry::Invalid();
+    }
+
+    return FirstServerClusterEntry(before.mEndpointId, endpoint, *cluster_idx + 1, mClusterIterationHint);
+}
+
+std::optional<InteractionModel::ClusterInfo> CodegenDataModel::GetClusterInfo(const ConcreteClusterPath & path)
+{
+    const EmberAfCluster * cluster = FindServerCluster(path);
+
+    VerifyOrReturnValue(cluster != nullptr, std::nullopt);
+
+    InteractionModel::ClusterInfo info;
+    LoadClusterInfo(path, *cluster, &info);
+
+    return std::make_optional(info);
+}
+
+InteractionModel::AttributeEntry CodegenDataModel::FirstAttribute(const ConcreteClusterPath & path)
+{
+    const EmberAfCluster * cluster = FindServerCluster(path);
+
+    VerifyOrReturnValue(cluster != nullptr, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributeCount > 0, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributes != nullptr, InteractionModel::AttributeEntry::Invalid());
+
+    return AttributeEntryFrom(path, cluster->attributes[0]);
+}
+
+std::optional<unsigned> CodegenDataModel::TryFindAttributeIndex(const EmberAfCluster * cluster, chip::AttributeId id) const
+{
+    const unsigned attributeCount = cluster->attributeCount;
+
+    // attempt to find this based on the embedded hint
+    if ((mAttributeIterationHint < attributeCount) && (cluster->attributes[mAttributeIterationHint].attributeId == id))
+    {
+        return std::make_optional(mAttributeIterationHint);
+    }
+
+    // linear search is required. This may be slow
+    for (unsigned attribute_idx = 0; attribute_idx < attributeCount; attribute_idx++)
+    {
+
+        if (cluster->attributes[attribute_idx].attributeId == id)
+        {
+            return std::make_optional(attribute_idx);
+        }
+    }
+
+    return std::nullopt;
+}
+
+const EmberAfCluster * CodegenDataModel::FindServerCluster(const ConcreteClusterPath & path)
+{
+    // cache things
+    if (mPreviouslyFoundCluster.has_value() && (mPreviouslyFoundCluster->path == path))
+    {
+        return mPreviouslyFoundCluster->cluster;
+    }
+
+    const EmberAfCluster * cluster = emberAfFindServerCluster(path.mEndpointId, path.mClusterId);
+    if (cluster != nullptr)
+    {
+        mPreviouslyFoundCluster = std::make_optional<ClusterReference>(path, cluster);
+    }
+    return cluster;
+}
+
+InteractionModel::AttributeEntry CodegenDataModel::NextAttribute(const ConcreteAttributePath & before)
+{
+    const EmberAfCluster * cluster = FindServerCluster(before);
+    VerifyOrReturnValue(cluster != nullptr, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributeCount > 0, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributes != nullptr, InteractionModel::AttributeEntry::Invalid());
+
+    // Handles global attribute iteration: if we got a global attribute, move to the next
```

Comment:
This is really fragile.  Can we reuse `GlobalAttributesNotInMetadata` here, somehow?

But even better would be for global list attributes to be handled at the level _above_ this API, not in here, so that every data model doesn't have to reimplement this part.

Reply:
Global attribute data needs to be returned and it seems like readattribute is the one providing it (so part of the data model). It seems consistent to make the data model both iterate and return data (and support writes as applicable) for all attributes rather than saying "these attributes will be iterated for you, however you do need to implement read/write). Separation of concerns in that case seems better.

Reply:
Removed global attribute support.